assignments/this.js

```diff
@@ -1,26 +1,59 @@
 /* The for principles of "this";
 * in your own words. explain the four principle for the "this" keyword below.
 *
-* 1. 
-* 2. 
-* 3. 
-* 4. 
+* 1. Global Context - Java code executes in global context. When 'this' is called in global scope it will refer to global object.
+*  If we call 'this' from function`s scope, it will refer to global object as well. In this two cases our context is global scope.
+* 2. Implicit Binding - This is used for adding the same statements to different context.
+*  We can just add the function to needed context (object) and then this keyword will refer to the context. In this case our context is object`s scope
+* 3. Explicit Binding - (call, apply, bind) -
+* 4. New Binding -
 *
 * write out a code example of each explanation above
 */
 
 // Principle 1
-
+this
 // code example for Window Binding
-
+function globalContext() {
+    console.log(this);
+    return 'gav';
+}
+globalContext();
 // Principle 2
-
 // code example for Implicit Binding
-
+const sayHi = (obj) => {
+  obj.sayHello = function () {
+      console.log(`My context is ${this.species}`);
+      return `Hello, I am a ${this.species}`;
+  };
+};
+const dog = {species: 'dog'};
+const cat = {species: 'cat'};
+sayHi(dog);
+sayHi(cat);
+console.log(dog.sayHello());
+console.log(cat.sayHello());
```

Comment:
correct as long as you know what "this" points to.  When 'this' is used in object methods then implicit binding is involved and 'this' is automaticaly assigned to an object when called by a preceding dot. So, by calling the objectName.method() the object name immediately to the left of the dot is the one attached to 'this'.  If a long list of objectnames: objectName1.objectName2.objectName3.method() - it is the last dot that is the one that counts.

// code example for Implicit Binding
const Chessplayer = {
    name: 'Fabiano',
    whiteOpening: 'Spanish Game',
    blackOpening: 'Sicilian',
    speak: function() {
        console.log(`My name is ${this.name} and my favorite openings are the ${this.whiteOpening} and the ${this.blackOpening}`);
    }
}

Chessplayer.speak();  // here, 'this' points to what is to the left of the dot - Chessplayer